storage/postgres/admin_storage.go

```diff
@@ -278,19 +351,112 @@ func (t *adminTX) CreateTree(ctx context.Context, tree *trillian.Tree) (*trillia
-	return nil, errNotImplemented
+	tree, err := t.GetTree(ctx, treeID)
+	if err != nil {
+		return nil, err
+	}
+
+	beforeUpdate := *tree
+	updateFunc(tree)
+	if err := storage.ValidateTreeForUpdate(ctx, &beforeUpdate, tree); err != nil {
+		return nil, err
+	}
+	if err := validateStorageSettings(tree); err != nil {
+		return nil, err
+	}
+
+	// Use the time truncated-to-millis throughout, as that's what's stored.
+	nowMillis := storage.ToMillisSinceEpoch(time.Now())
+	now := storage.FromMillisSinceEpoch(nowMillis)
+	tree.UpdateTime, err = ptypes.TimestampProto(now)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build tree.UpdateTime: %v", err)
+	}
+	rootDuration, err := ptypes.Duration(tree.MaxRootDuration)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse MaxRootDuration: %v", err)
+	}
+
+	privateKey, err := proto.Marshal(tree.PrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("could not marshal PrivateKey: %v", err)
+	}
+
+	stmt, err := t.tx.PrepareContext(ctx, updateTreeSQL)
+	if err != nil {
+		return nil, err
+	}
+	defer stmt.Close()
+
+	if _, err = stmt.ExecContext(
+		ctx,
+		tree.TreeState.String(),
+		tree.TreeType.String(),
+		tree.DisplayName,
+		tree.Description,
+		nowMillis,
+		rootDuration/time.Millisecond,
+		privateKey,
+		tree.TreeId); err != nil {
+		return nil, err
+	}
+
+	return tree, nil
 }
 
 func (t *adminTX) SoftDeleteTree(ctx context.Context, treeID int64) (*trillian.Tree, error) {
-	return nil, errNotImplemented
+	return t.updateDeleted(ctx, treeID, true /* deleted */, storage.ToMillisSinceEpoch(time.Now()) /* deleteTimeMillis */)
 }
 
 func (t *adminTX) UndeleteTree(ctx context.Context, treeID int64) (*trillian.Tree, error) {
-	return nil, errNotImplemented
+	return t.updateDeleted(ctx, treeID, false /* deleted */, nil /* deleteTimeMillis */)
 }
 
 func (t *adminTX) HardDeleteTree(ctx context.Context, treeID int64) error {
-	return errNotImplemented
+	if err := validateDeleted(ctx, t.tx, treeID, true /* wantDeleted */); err != nil {
+		return err
+	}
+
+	if _, err := t.tx.ExecContext(ctx, deleteFromTreeControlSQL, treeID); err != nil {
+		return err
+	}
+	_, err := t.tx.ExecContext(ctx, deleteFromTreesSQL, treeID)
+	return err
+}
+
+// updateDeleted updates the Deleted and DeleteTimeMillis fields of the specified tree.
+// deleteTimeMillis must be either an int64 (in millis since epoch) or nil.
+func (t *adminTX) updateDeleted(ctx context.Context, treeID int64, deleted bool, deleteTimeMillis interface{}) (*trillian.Tree, error) {
+	if err := validateDeleted(ctx, t.tx, treeID, !deleted); err != nil {
+		return nil, err
+	}
+	if _, err := t.tx.ExecContext(
+		ctx,
+		softDeleteSQL,
+		deleted,
+		deleteTimeMillis,
+		treeID); err != nil {
+		return nil, err
+	}
+	return t.GetTree(ctx, treeID)
+}
+
+func validateDeleted(ctx context.Context, tx *sql.Tx, treeID int64, wantDeleted bool) error {
+	var nullDeleted sql.NullBool
```

Comment:
The `deleted` column is no longer nullable, so this code can be simplified.

Reply:
Sorry for all the confusion, I appreciate the patience